rapl-configure/rapl-configure.1.in
.TH "rapl-configure-@RAPL_LIB@" "1" "2024-01-05" "RAPLCap @PROJECT_VERSION@" "RAPLCap Utilities"
.SH "NAME"
.LP
rapl\-configure\-@RAPL_LIB@ \- get or set Intel RAPL configurations
.SH "SYNPOSIS"
.LP
\fBrapl\-configure\-@RAPL_LIB@\fP [\fIOPTION\fP]...
.SH "DESCRIPTION"
.LP
Intel Running Average Power Limit (RAPL) configures \fIpower limits\fP on
hardware that are respected on average over a \fItime window\fP.
An Intel\*R Sandy Bridge (2nd generation Intel\*R Core) or newer processor is
required.
Users are expected to have some familiarity with RAPL before using this tool.
Refer to the "Intel\*R 64 and IA-32 Architectures Software Developer Manual,
Volume 3: System Programming Guide" for more technical information.
This description is only a short summary.
.LP
RAPL control is structured into \fIzones\fP, also known as \fIdomains\fP.
Currently supported zones are:
.IP \[bu] 2
\fIPackage\fP: a processor package die, including cores and other shared
resources like the uncore components below and a DRAM controller.
.IP \[bu]
\fICore\fP: the processor core, like compute components and private caches
(often L1 and L2).
Core is also known as \fIPP0\fP, or Power Plane 0.
.IP \[bu]
\fIUncore\fP: non-core processor components like shared/last-level caches
(often L3) and any integrated graphics hardware.
Uncore is also known as \fIPP1\fP, or Power Plane 1, and is typically only
available on client processors like those for desktops and laptops.
.IP \[bu]
\fIDRAM\fP: directly-attached main memory.
DRAM power capping is typically only available on server systems, but has been
spotted on some client hardware.
.IP \[bu]
\fIPSys\fP: the entire System-on-a-Chip (SoC).
PSys is also known as \fIPlatform\fP and is only available on Intel\*R Skylake
(6th generation Intel\*R Core) and newer processors, and requires additional
BIOS support.
.LP
Package and PSys zones support two power limits, or \fIconstraints\fP - \fIlong
term\fP and \fIshort term\fP.
Package zones may also support a \fIpeak power\fP constraint on Tiger Lake
(11th generation Intel\*R Core) & newer client processors.
Other zones support a single constraint (long term).
.LP
Actual power limit and time window values are system-dependent.
Refer to your hardware's technical documentation, particularly for the Thermal
Design Power (TDP).
.LP
Package, Core, Uncore, and DRAM zones may exist for each die in each processor
package, although many systems only have a single package with a single die.
.LP
By default, rapl\-configure\-@RAPL_LIB@ uses the PACKAGE zone on package 0,
die 0.
Current values are printed if no flags, or only package, die, and/or zone flags
are specified.
Otherwise, specified values are set while other values remain unmodified.
.SH "OPTIONS"
.LP
.TP
\fB\-h,\fP \fB\-\-help\fP
Print the help screen and exit
.TP
\fB\-n,\fP \fB\-\-npackages\fP
Print the number of packages found and exit
.TP
\fB\-N,\fP \fB\-\-ndie\fP
Print the number of die found for a package and exit
.TP
\fB\-c,\fP \fB\-\-package\fP=\fIPACKAGE\fP
The processor package (0 by default)
.TP
\fB\-d,\fP \fB\-\-die\fP=\fIDIE\fP
The package die (0 by default)
.TP
\fB\-z,\fP \fB\-\-zone\fP=\fIZONE\fP
Which zone/domain use. Allowable values:
.br
PACKAGE \- a processor package (default)
.br
CORE \- core power plane
.br
UNCORE \- uncore power plane
.br
DRAM \- main memory
.br
PSYS \- the entire SoC
.TP
\fB\-l,\fP \fB\-\-limit\fP=\fICONSTRAINT\fP
Which limit/constraint to use. Allowable values:
.br
LONG \- long term (default)
.br
SHORT \- short term
.br
PEAK \- peak power
.TP
\fB\-t,\fP \fB\-\-time\fP=\fISECONDS\fP
Constraint's time window
.TP
\fB\-p,\fP \fB\-\-power\fP=\fIWATTS\fP
Constraint's power limit
.TP
\fB\-e,\fP \fB\-\-enabled\fP=\fI1|0\fP
Enable/disable a zone
.LP
The following allow setting long and short term constraints simultaneously:
.TP
\fB\-s,\fP \fB\-\-seconds0\fP=\fISECONDS\fP
Long term time window
.TP
\fB\-w,\fP \fB\-\-watts0\fP=\fIWATTS\fP
Long term power limit
.TP
\fB\-S,\fP \fB\-\-seconds1\fP=\fISECONDS\fP
Short term time window
.TP
\fB\-W,\fP \fB\-\-watts1\fP=\fIWATTS\fP
Short term power limit
.SH "EXAMPLES"
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-n\fP
Print the number of packages found on the system.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-N \-c 1\fP
Print the number of die found for package 1 (requires multi-package system).
.TP
\fBrapl\-configure\-@RAPL_LIB@\fP
Print PACKAGE zone information for package 0, die 0.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-d 1\fP
Print PACKAGE zone information for package 0, die 1 (requires multi-die
package).
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-c 1\fP
Print PACKAGE zone information for package 1, die 0 (requires multi-package
system).
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-z CORE\fP
Print CORE zone information for package 0, die 0.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-c 1 \-z DRAM\fP
Print DRAM zone information for package 1, die 0 (requires multi-package
system).
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-p 20\fP
Set 20 Watt long term power constraint on PACKAGE zone for package 0, die 1
without changing time window.
The \fB\-w\fP option would work in place of \fB\-p\fP for this constraint.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-p 20 \-t 30\fP
Set 20 Watt long term power constraint with 30 second time window on PACKAGE
zone for package 0, die 0.
The \fB\-w\fP and \fB\-s\fP options would work in place of \fB\-p\fP and
\fB\-t\fP, respectively, for this constraint.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-l SHORT \-p 30 \-t 0.0025\fP
Set 30 Watt short term power constraint with 2.5 millisecond time window on
PACKAGE zone for package 0, die 0.
The \fB\-W\fP and \fB\-S\fP options would work in place of \fB\-p\fP and
\fB\-t\fP, respectively, for this constraint, in which case \fB\-l SHORT\fP
would not be needed.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-w 20 \-s 30 \-W 30 \-S 0.0025\fP
Set 20 Watt long term power constraint with 30 second time window and
set 30 Watt short term power constraint with 2.5 millisecond time window on
PACKAGE zone for package 0, die 0.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-l PEAK \-p 45\fP
Set 45 Watt peak power constraint on PACKAGE zone for package 0, die 0.
Peak power does not accept a time window value.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-z UNCORE \-p 10 \-e 1\fP
Set 10 Watt long term power constraint on UNCORE zone for package 0, die 0
without changing time window, and enable the zone.
The \fB\-w\fP option would work in place of \fB\-p\fP for this constraint.
.TP
\fBrapl\-configure\-@RAPL_LIB@ \-z UNCORE \-e 0\fP
Disable UNCORE zone for package 0, die 0.
.SH "REMARKS"
.LP
Administrative (root) privileges are usually needed to access RAPL settings.
.LP
Not all zones are supported by all systems.
Some systems may not allow configuring or enabling/disabling supported zones,
even if they expose the configurations.
.LP
Power and time window values are not always stored exactly as specified.
This is a result of how values are encoded in the Model-Specific Registers.
When using the powercap backend, some accuracy may also be lost as the
interface forces rounding values to microJoule and microsecond precision.
.LP
Historically, the term \fIsocket\fP was used interchangeably with \fIpackage\fP.
While there used to only ever be a single processor package per socket, this
symmetry no longer holds - sockets may contain multiple packages.
.SH "BUGS"
.LP
Report bugs upstream at <https://github.com/powercap/raplcap>
.SH "FILES"
.nf
\fI/dev/cpu/*/msr\fP
\fI/sys/class/powercap/intel\-rapl:*/\fP
.fi

RELEASES.md
# Release Notes

## [Unreleased]

### Added

* [msr] Support for Emerald Rapids server ("Raptor Cove"), Model 0xCF
* [msr] Support for Meteor Lake mobile ("Redwood Cove" / "Crestmont"), Model 0xAA

### Changed

* [msr] Updated for the Intel Software Developer's Manual, December 2023 release
* [rapl-configure] Refactor help output and man page documentation on zones and constraints

### Removed

* [ipg] Remove experimental Intel Power Gadget backend (no longer supported by Microsoft)

### Fixed

* [msr] Incorrect maximum power limit for Sapphire Rapids PSYS zone


## [v0.9.1] - 2023-11-19

### Changed

* [msr] Updated for the Intel Software Developer's Manual, September 2023 release

### Fixed

* [msr] Support check misses some CPUs: Raptor Lake, Sapphire Rapids, and SoFIA ([#14])
* [powercap] Peak power constraint shouldn't look for a time window ([#14])


## [v0.9.0] - 2023-04-30

### Added

* [msr] Support for Raptor Lake ("Raptor Cove" / "Gracemont"), Model 0xB7
* [msr] Support for Raptor Lake mobile ("Raptor Cove" / "Gracemont"), Model 0xBA
* [msr] Support for Raptor Lake client ("Raptor Cove" / "Gracemont"), Model 0xBF
* [msr] Support for Sapphire Rapids server ("Golden Cove"), Model 0x8F

### Changed

* [msr] Updated for the Intel Software Developer's Manual, March 2023 release

### Fixed

* pkg-config file is broken when CMAKE_INSTALL_{INCLUDE,LIB}DIR is absolute


## [v0.8.0] - 2022-05-13

### Added

* [msr] Support for Atom Tremont client ("Elkhart Lake"), Model 0x96
* [msr] Support for Atom Tremont mobile ("Jasper Lake"), Model 0x9C
* [msr] Support for Alder Lake client ("Golden Cove" / "Gracemont"), Model 0x97
* [msr] Support for Alder Lake mobile ("Golden Cove" / "Gracemont"), Model 0x9A

### Changed

* [msr] Updated for the Intel Software Developer's Manual, April 2022 release
* Changes based on previously deprecated behavior:
  * [powercap] Don't fall back on PSYS zone at die=0 and then at pkg=0 when not found for specified pkg/die
* Internal tweaks based on testing with additional compiler flags

### Fixed

* [msr] Intel SDM incorrectly documents energy status units for Atom Silvermont micro server, Model 0x4D ([#13])


## [v0.7.0] - 2021-12-15

### Added

* CMake package configs
* CMake installation components
* [msr] Support for Atom Silvermont micro server ("Avoton" / "Rangeley"), Model 0x4D
* [powercap] Support for multiple PSYS zones (Linux kernel 5.14 and newer, e.g., for Sapphire Rapids CPUs)

### Changed

* Increased minimum CMake version from 3.6 to 3.12
* Changes based on previously deprecated behavior:
  * [rapl-configure] Print fully-qualified "watts_long" and "seconds_long" in output for zones with a single constraint
  * [rapl-configure] Don't automatically enable zones - user must explicitly set -e/--enabled=1
  * [rapl-configure] Fail if -W/--watts1 and/or -S/--seconds1 are used but short term constraint is not supported

### Deprecated

* [powercap] Falling back on PSYS zone at die=0 and then at pkg=0 when not found for specified pkg/die - will return an error in the future

### Fixed

* [powercap] Memory leak in 'raplcap_destroy'
* [rapl-configure] Missing error check when querying for peak power (PL4) constraint support


## [v0.6.0] - 2021-05-12

### Added

* Support for constraint-specific operations
  * Interface enum 'raplcap_constraint'
  * Interface function 'raplcap_pd_is_constraint_supported'
  * Interface function 'raplcap_pd_get_limit'
  * Interface function 'raplcap_pd_set_limit'
  * [msr] Interface function 'raplcap_msr_pd_is_locked'
  * [msr] Interface function 'raplcap_msr_pd_set_locked'
  * [rapl-configure] New -l/--limit option
  * [rapl-configure] New -t/--time option
  * [rapl-configure] New -p/--power option
* Support for peak power (PL4) constraint
* [msr] Support for Tiger Lake, Models 0x8C, 0x8D

### Changed

* Increased minimum CMake version from 2.8.12 to 3.6
* [msr] Updated for the Intel Software Developer's Manual, November 2020 release

### Deprecated

* Add compiler deprecation attributes to deprecated functions and fields
* [rapl-configure] Automatic enabling of zones (without an explicit -e/--enabled=1 argument)
* [rapl-configure] Printing "watts" and "seconds" in output for zones with a single constraint, will use fully-qualified names in the future
* [rapl-configure] Ignoring -S/--seconds1 and -W/--watts1 options for zones without short term constraints, will be an error in the future

### Removed

* [rapl-configure] Documentation on previously deprecated long options

### Fixed

* [msr] Writing to wrong die MSR in 'raplcap_pd_set_limits', 'raplcap_msr_pd_set_zone_clamped', and 'raplcap_msr_pd_set_zone_locked'
* [powercap] Library's pkg-config file should specify the minimum powercap dependency version
* [rapl-configure] Incorrect man page example for disabling a zone


## [v0.5.0] - 2020-09-02

### Added

* Support for multi-die packages ([#9])
  * Interface function 'raplcap_get_num_die'
  * Interface functions 'raplcap_pd_*'
  * [msr] Interface functions 'raplcap_msr_pd_*'
  * [rapl-configure] New -N/--ndie flag
  * [rapl-configure] New -d/--die option
* Interface function 'raplcap_get_num_packages' (supersedes 'raplcap_get_num_sockets')
* [msr] Support for Ice Lake, Models 0x7D, 0x7E, 0x6A, 0x6C
* [msr] Support for Comet Lake, Models 0xA5, 0xA6
* [rapl-configure] Long option --npackages (supersedes --nsockets)
* [rapl-configure] Long option --package (supersedes --socket)

### Changed

* Increased minimum CMake version from 2.8.5 to 2.8.12 to support target_compile_definitions
* [msr] Updated for the Intel Software Developer's Manual, May 2020 release
* [powercap] Library dependency now requires version >= 0.3.0

### Deprecated

* Functions that are superseded by die-aware ones ([#9])
  * Interface functions that are not die-aware, use corresponding 'raplcap_pd_*' functions instead
  * [msr] Interface functions that are not die-aware, use corresponding 'raplcap_msr_pd_*' functions instead

rapl-configure/rapl-configure.c
/**
 * Get/set RAPL values.
 *
 * @author Connor Imes
 * @date 2016-05-13
 */
// for setenv
#define _POSIX_C_SOURCE 200112L
#include <assert.h>
#include <errno.h>
#include <stdio.h>
#include <stdlib.h>
#include <string.h>
#include <getopt.h>
#include "raplcap.h"
#include "raplcap-common.h"
#ifdef RAPLCAP_msr
#include "raplcap-msr.h"
#endif // RAPLCAP_msr

typedef struct rapl_configure_ctx {
  int get_packages;
  int get_die;
  raplcap_zone zone;
  raplcap_constraint constraint;
  unsigned int pkg;
  unsigned int die;
  int enabled;
  int set_enabled;
  int set_long;
  raplcap_limit limit_long;
  int set_short;
  raplcap_limit limit_short;
  int set_constraint;
  raplcap_limit limit_constraint;
#ifdef RAPLCAP_msr
  int clamped;
  int set_clamped;
  int set_locked;
#endif // RAPLCAP_msr
} rapl_configure_ctx;

static const char* prog;
static const char short_options[] = "nNc:d:z:l:t:p:e:s:w:S:W:C:Lh";
static const struct option long_options[] = {
  {"npackages",no_argument,       NULL, 'n'},
  {"nsockets", no_argument,       NULL, 'n'}, // deprecated, no longer documented
  {"ndie",     no_argument,       NULL, 'N'},
  {"package",  required_argument, NULL, 'c'},
  {"socket",   required_argument, NULL, 'c'}, // deprecated, no longer documented
  {"die",      required_argument, NULL, 'd'},
  {"zone",     required_argument, NULL, 'z'},
  {"limit",    required_argument, NULL, 'l'},
  {"time",     required_argument, NULL, 't'},
  {"power",    required_argument, NULL, 'p'},
  {"enabled",  required_argument, NULL, 'e'},
  {"seconds0", required_argument, NULL, 's'},
  {"watts0",   required_argument, NULL, 'w'},
  {"seconds1", required_argument, NULL, 'S'},
  {"watts1",   required_argument, NULL, 'W'},
#ifdef RAPLCAP_msr
  {"clamped",  required_argument, NULL, 'C'},
  {"locked",   no_argument,       NULL, 'L'},
#endif // RAPLCAP_msr
  {"help",     no_argument,       NULL, 'h'},
  {0, 0, 0, 0}
};

__attribute__ ((noreturn))
static void print_usage(int exit_code) {
  fprintf(exit_code ? stderr : stdout,
          "Usage: %s [OPTION]...\n"
          "Options:\n"
          "  -h, --help               Print this message and exit\n"
          "  -n, --npackages          Print the number of packages found and exit\n"
          "  -N, --ndie               Print the number of die found for a package and exit\n"
          "  -c, --package=PACKAGE    The processor package (0 by default)\n"
          "  -d, --die=DIE            The package die (0 by default)\n"
          "  -z, --zone=ZONE          Which zone/domain use. Allowable values:\n"
          "                           PACKAGE - a processor package (default)\n"
          "                           CORE - core power plane\n"
          "                           UNCORE - uncore power plane\n"
          "                           DRAM - main memory\n"
          "                           PSYS - the entire SoC\n"
          "  -l, --limit=CONSTRAINT   Which limit/constraint to use. Allowable values:\n"
          "                           LONG - long term (default)\n"
          "                           SHORT - short term\n"
          "                           PEAK - peak power\n"
          "  -t, --time=SECONDS       Constraint's time window\n"
          "  -p, --power=WATTS        Constraint's power limit\n"
          "  -e, --enabled=1|0        Enable/disable a zone\n"
#ifdef RAPLCAP_msr
          "  -C, --clamped=1|0        Clamp/unclamp a zone\n"
          "                           Clamping is automatically set when enabling\n"
          "                           Therefore, you MUST explicitly set --clamped=0 when\n"
          "                           setting limits (since zones are auto-enabled)\n"
          "  -L, --locked             Lock a zone (a core RESET is required to unlock)\n"
#endif // RAPLCAP_msr
          "The following allow setting long and short term constraints simultaneously:\n"
          "  -s, --seconds0=SECONDS   Long term time window\n"
          "  -w, --watts0=WATTS       Long term power limit\n"
          "  -S, --seconds1=SECONDS   Short term time window\n"
          "  -W, --watts1=WATTS       Short term power limit\n"
          "\nCurrent values are printed if no flags, or only package, die, and/or zone flags are specified.\n"
          "Otherwise, specified values are set while other values remain unmodified.\n"
          "\nRAPL is available on Intel CPUs starting with Sandy Bridge (2011).\n"
          "Configurability varies by system, but the following are informal guidelines.\n"
          "ZONE support:\n"
          "  PACKAGE: all CPUs\n"
          "  CORE: all CPUs, except maybe some Atom CPUs\n"
          "  UNCORE: client CPUs\n"
          "  DRAM: server CPUs\n"
          "  PSYS: Skylake (2015) & newer CPUs\n"
          "CONSTRAINT support:\n"
          "  LONG: all zones\n"
          "  SHORT: PACKAGE & PSYS zones\n"
          "  PEAK: PACKAGE zone on Tiger Lake (2020) & newer client CPUs\n",
          prog);
  exit(exit_code);
}

// something reasonably outside of errno range
#define PRINT_LIMIT_IGNORE -1000

static void print_limits(int enabled, int locked, int clamped,
                         double watts_long, double seconds_long,
                         double watts_short, double seconds_short,
                         int locked_peak, double watts_peak,
                         double joules, double joules_max) {
  // Note: simply using %f (6 decimal places) doesn't provide sufficient precision
  const char* en = enabled < 0 ? "unknown" : (enabled ? "true" : "false");
  const char* lck = locked < 0 ? "unknown" : (locked ? "true" : "false");
  const char* clmp = clamped < 0 ? "unknown" : (clamped ? "true" : "false");
  // time window can never be 0, so if it's > 0, the short term constraint exists
  printf("%13s: %s\n", "enabled", en);
  if (clamped != PRINT_LIMIT_IGNORE) {
    printf("%13s: %s\n", "clamped", clmp);
  }
  if (locked != PRINT_LIMIT_IGNORE) {
    printf("%13s: %s\n", "locked", lck);
  }
  printf("%13s: %.12f\n", "watts_long", watts_long);
  printf("%13s: %.12f\n", "seconds_long", seconds_long);
  if (seconds_short > 0) {
    printf("%13s: %.12f\n", "watts_short", watts_short);
    printf("%13s: %.12f\n", "seconds_short", seconds_short);
  }
  if (locked_peak != PRINT_LIMIT_IGNORE) {
    printf("%13s: %s\n", "locked_peak", (locked_peak ? "true" : "false"));
  }
  if (watts_peak > 0) {
    printf("%13s: %.12f\n", "watts_peak", watts_peak);
  }
  if (joules >= 0) {
    printf("%13s: %.12f\n", "joules", joules);
  }
  if (joules_max >= 0) {
    printf("%13s: %.12f\n", "joules_max", joules_max);
  }
}

static void print_error_continue(const char* msg) {
  perror(msg);
  fprintf(stderr, "Trying to proceed anyway...\n");
}

static int configure_limits(const rapl_configure_ctx* c) {
  assert(c != NULL);
  int ret;
  // set limits
  if ((c->set_long || c->set_short)) {
    if ((ret = raplcap_pd_set_limits(NULL, c->pkg, c->die, c->zone,
                                     c->set_long ? &c->limit_long : NULL,
                                     c->set_short ? &c->limit_short : NULL))) {
      perror("Failed to set limits");
      return ret;
    }
  }
  if (c->set_constraint) {
    if ((ret = raplcap_pd_set_limit(NULL, c->pkg, c->die, c->zone, c->constraint, &c->limit_constraint))) {
      perror("Failed to set limit");
      return ret;
    }
  }

  if (c->set_enabled) {
    if ((ret = raplcap_pd_set_zone_enabled(NULL, c->pkg, c->die, c->zone, c->enabled))) {
      perror("Failed to enable/disable zone");
      return ret;
    }
  }

#ifdef RAPLCAP_msr
  // Note: Enabling automatically sets clamping AND we auto-enable when configuring unless explicitly requested not to.
  //       As a result:
  //       1) We set clamping here AFTER enabling in case clamping was requested to be off
  //       2) The user must always explicitly request clamping to be off when setting RAPL limits
  // No clamped field for peak power
  if (c->set_clamped) {
    if (c->set_long || c->set_short ||
        c->constraint == RAPLCAP_CONSTRAINT_LONG_TERM || c->constraint == RAPLCAP_CONSTRAINT_SHORT_TERM) {
      if ((ret = raplcap_msr_pd_set_zone_clamped(NULL, c->pkg, c->die, c->zone, c->clamped))) {
        perror("Failed to clamp/unclamp zone");
        return ret;
      }
    }
  }

  if (c->set_locked) {
    if (c->set_long || c->set_short ||
        c->constraint == RAPLCAP_CONSTRAINT_LONG_TERM || c->constraint == RAPLCAP_CONSTRAINT_SHORT_TERM) {
      if ((ret = raplcap_msr_pd_set_zone_locked(NULL, c->pkg, c->die, c->zone))) {
        perror("Failed to lock zone");
        return ret;
      }
    }
    // Intentionally not an "else if" - user can set long/short term using -s/-w/-S/-W and also use -l for peak power
    // This if statement isn't strictly necessary, but prevents setting locked twice if constraint is long or short term
    if (c->constraint == RAPLCAP_CONSTRAINT_PEAK_POWER) {
      if ((ret = raplcap_msr_pd_set_locked(NULL, c->pkg, c->die, c->zone, c->constraint))) {
        perror("Failed to lock peak power");
        return ret;
      }
    }
  }
#endif // RAPLCAP_msr
  return 0;
}

static int get_limits(unsigned int pkg, unsigned int die, raplcap_zone zone) {
  raplcap_limit ll = { 0 };
  raplcap_limit ls = { 0 };
  raplcap_limit lp = { 0 };
  double joules;
  double joules_max;
  int supp;
  int locked = PRINT_LIMIT_IGNORE;
  int clamped = PRINT_LIMIT_IGNORE;
  int locked_peak = PRINT_LIMIT_IGNORE;
  int ret;
  int enabled = raplcap_pd_is_zone_enabled(NULL, pkg, die, zone);
  if (enabled < 0) {
    print_error_continue("Failed to determine if zone is enabled");
  }
#ifdef RAPLCAP_msr
  locked = raplcap_msr_pd_is_zone_locked(NULL, pkg, die, zone);
  if (locked < 0) {
    print_error_continue("Failed to determine if zone is locked");
  }
  clamped = raplcap_msr_pd_is_zone_clamped(NULL, pkg, die, zone);
  if (clamped < 0) {
    print_error_continue("Failed to determine if zone is clamped");
  }
#endif // RAPLCAP_msr
  if ((ret = raplcap_pd_get_limits(NULL, pkg, die, zone, &ll, &ls))) {
    perror("Failed to get limits");
    return ret;
  }
  supp = raplcap_pd_is_constraint_supported(NULL, pkg, die, zone, RAPLCAP_CONSTRAINT_PEAK_POWER);
  if (supp < 0) {
    print_error_continue("Failed to determine if peak power constraint is supported");
  } else if (supp) {
#ifdef RAPLCAP_msr
    locked_peak = raplcap_msr_pd_is_locked(NULL, pkg, die, zone, RAPLCAP_CONSTRAINT_PEAK_POWER);
    if (locked < 0) {
      print_error_continue("Failed to determine if peak power is locked");
    }
#endif // RAPLCAP_msr
    if ((ret = raplcap_pd_get_limit(NULL, pkg, die, zone, RAPLCAP_CONSTRAINT_PEAK_POWER, &lp))) {
      perror("Failed to get peak power limit");
      return ret;
    }
  }
  // we'll consider energy counter information to be optional
  joules = raplcap_pd_get_energy_counter(NULL, pkg, die, zone);
  joules_max = raplcap_pd_get_energy_counter_max(NULL, pkg, die, zone);
  print_limits(enabled, locked, clamped,
               ll.watts, ll.seconds, ls.watts, ls.seconds,
               locked_peak, lp.watts,
               joules, joules_max);
  return ret;
}

static int check_zone_supported(const rapl_configure_ctx* c) {
  assert(c != NULL);
  int supported = raplcap_pd_is_zone_supported(NULL, c->pkg, c->die, c->zone);
  if (supported < 0) {
    print_error_continue("Failed to determine if zone is supported");
  } else if (supported == 0) {
    fprintf(stderr, "Zone not supported\n");
    return -1;
  }
  return 0;
}

static int check_constraint_supported(const rapl_configure_ctx* c) {
  assert(c != NULL);
  // Check the -l constraint
  int supported = raplcap_pd_is_constraint_supported(NULL, c->pkg, c->die, c->zone, c->constraint);
  if (supported < 0) {
    print_error_continue("Failed to determine if constraint is supported");
  } else if (supported == 0) {
    fprintf(stderr, "Constraint not supported\n");
    return -1;
  }
  // Check that time window is not specified for peak power constraint
  if (c->constraint == RAPLCAP_CONSTRAINT_PEAK_POWER && c->limit_constraint.seconds > 0) {
    fprintf(stderr, "Cannot set a time window for peak power\n");
    return -1;
  }
  // Check short term constraint if -W and/or -S are used
  // All zones (at least currently) support long term constraints, so no need to check c->set_long
  if (c->set_short) {
    supported = raplcap_pd_is_constraint_supported(NULL, c->pkg, c->die, c->zone, RAPLCAP_CONSTRAINT_SHORT_TERM);
    if (supported < 0) {
      print_error_continue("Failed to determine if short term constraint is supported");
    } else if (supported == 0) {
      fprintf(stderr, "Short term constraint not supported\n");
      return -1;
    }
  }
  return 0;
}

#define SET_VAL(optarg, val, set_val) \
  if ((val = atof(optarg)) <= 0) { \
    fprintf(stderr, "Time window and power limit values must be > 0\n"); \
    print_usage(1); \
  } \
  set_val = 1

int main(int argc, char** argv) {
  rapl_configure_ctx ctx = { 0 };
  int ret = 0;
  int c;
  uint32_t count;
  prog = argv[0];
  int is_read_only;

  // parse parameters
  while ((c = getopt_long(argc, argv, short_options, long_options, NULL)) != -1) {
    switch (c) {
      case 'h':
        print_usage(0);
        break;
      case 'c':
        ctx.pkg = (unsigned int) atoi(optarg);
        break;
      case 'd':
        ctx.die = (unsigned int) atoi(optarg);
        break;
      case 'n':
        ctx.get_packages = 1;
        break;
      case 'N':
        ctx.get_die = 1;
        break;
      case 'z':
        if (!strcmp(optarg, "PACKAGE")) {
          ctx.zone = RAPLCAP_ZONE_PACKAGE;
        } else if (!strcmp(optarg, "CORE")) {
          ctx.zone = RAPLCAP_ZONE_CORE;
        } else if (!strcmp(optarg, "UNCORE")) {
          ctx.zone = RAPLCAP_ZONE_UNCORE;
        } else if (!strcmp(optarg, "DRAM")) {
          ctx.zone = RAPLCAP_ZONE_DRAM;
        } else if (!strcmp(optarg, "PSYS")) {
          ctx.zone = RAPLCAP_ZONE_PSYS;
        } else {
          print_usage(1);
        }
        break;
      case 'l':
        if (!strcmp(optarg, "LONG")) {
          ctx.constraint = RAPLCAP_CONSTRAINT_LONG_TERM;
        } else if (!strcmp(optarg, "SHORT")) {
          ctx.constraint = RAPLCAP_CONSTRAINT_SHORT_TERM;
        } else if (!strcmp(optarg, "PEAK")) {
          ctx.constraint = RAPLCAP_CONSTRAINT_PEAK_POWER;
        } else {
          print_usage(1);
        }
        break;
      case 't':
        SET_VAL(optarg, ctx.limit_constraint.seconds, ctx.set_constraint);
        break;
      case 'p':
        SET_VAL(optarg, ctx.limit_constraint.watts, ctx.set_constraint);
        break;
      case 'e':
        ctx.enabled = atoi(optarg);
        ctx.set_enabled = 1;
        break;
      case 's':
        SET_VAL(optarg, ctx.limit_long.seconds, ctx.set_long);
        break;
      case 'w':
        SET_VAL(optarg, ctx.limit_long.watts, ctx.set_long);
        break;
      case 'S':
        SET_VAL(optarg, ctx.limit_short.seconds, ctx.set_short);
        break;
      case 'W':
        SET_VAL(optarg, ctx.limit_short.watts, ctx.set_short);
        break;
#ifdef RAPLCAP_msr
      case 'C':
        ctx.clamped = atoi(optarg);
        ctx.set_clamped = 1;
        break;
      case 'L':
        ctx.set_locked = 1;
        break;
#endif // RAPLCAP_msr
      case '?':
      default:
        print_usage(1);
        break;
    }
  }

  // just print the number of packages or die and exit
  // these are often unprivileged operations since we don't need to initialize a raplcap instance
  if (ctx.get_packages) {
    count = raplcap_get_num_packages(NULL);
    if (count == 0) {
      perror("Failed to get number of packages");
      return 1;
    }
    printf("%"PRIu32"\n", count);
    return 0;
  }
  if (ctx.get_die) {
    count = raplcap_get_num_die(NULL, ctx.pkg);
    if (count == 0) {
      perror("Failed to get number of die");
      return 1;
    }
    printf("%"PRIu32"\n", count);
    return 0;
  }

  // initialize
  is_read_only = !ctx.set_enabled && !ctx.set_long && !ctx.set_short && !ctx.set_constraint;
#ifdef RAPLCAP_msr
  is_read_only &= !ctx.set_clamped && !ctx.set_locked;
#endif // RAPLCAP_msr
#ifndef _WIN32
  if (is_read_only) {
    // request read-only access (not supported by all implementations, therefore not guaranteed)
    setenv(ENV_RAPLCAP_READ_ONLY, "1", 0);
  }
#endif
  if (raplcap_init(NULL)) {
    perror("Failed to initialize");
    return 1;
  }

  if (!(ret = check_zone_supported(&ctx))) {
    // perform requested action
    if (is_read_only) {
      // TODO: Should we limit output by constraint, too?
      ret = get_limits(ctx.pkg, ctx.die, ctx.zone);
    } else if (!(ret = check_constraint_supported(&ctx))) {
      ret = configure_limits(&ctx);
    }
  }

  // cleanup
  if (raplcap_destroy(NULL)) {
    perror("Failed to clean up");
  }

  return ret;
}
